### PR TITLE
EN 975 add categories

### PR DESCRIPTION
**Pull request by mcarta-rt**

### CODE REVIEW

**Comment by entando-service** [comment]
All is well

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "designtime-admin-app",
```
Comment (fedelocci) [inline package.json]: change in app-builder

File: src/api/categories.js [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
-import { CATEGORY_TREE } from 'test/mocks/categories';
+import { CATEGORY_TREE, BODY_OK } from 'test/mocks/categories';
 
-const getCategoryTreeMockResponse = (queryString) => {
-  switch (queryString) {
-    case '?parentNode=home': { return CATEGORY_TREE.home; }
-    case '?parentNode=mycategory1': { return CATEGORY_TREE.mycategory1; }
+const getCategoryTreeMockResponse = (categoryCode) => {
+  switch (categoryCode) {
+    case 'home': { return CATEGORY_TREE.home; }
```
Comment (fedelocci) [inline src/api/categories.js]: i will prefer using constant inside in the case

File: test/ui/categories/add/AddCategoryPage.test.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  let component;
+  beforeEach(() => {
+    component = shallow(<AddCategoryPage />);
+  });
+  it('renders without crashing', () => {
+    expect(component.exists()).toEqual(true);
+  });
+
+  it('verify if exist InternalPage with class AddCategoryPage', () => {
+    expect(component.find('InternalPage').hasClass('AddCategoryPage')).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/add/AddCategoryPage.test.js]: toBe

File: test/ui/categories/add/AddCategoryPage.test.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import { shallow } from 'enzyme';
+import AddCategoryPage from 'ui/categories/add/AddCategoryPage';
+
+describe('AddCategoryPage', () => {
+  let component;
+  beforeEach(() => {
+    component = shallow(<AddCategoryPage />);
+  });
+  it('renders without crashing', () => {
+    expect(component.exists()).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/add/AddCategoryPage.test.js]: toBe

File: test/ui/categories/add/AddFormContainer.test.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  toggleLoading: jest.fn(),
+}));
+
+describe('AddFormContainer', () => {
+  const dispatchMock = jest.fn();
+
+  describe('mapStateToProps', () => {
+    let props;
+    beforeEach(() => {
+      props = mapStateToProps({
```
Comment (fedelocci) [inline test/ui/categories/add/AddFormContainer.test.js]: it's prefered using a empty state. you don't test the selectors

File: test/ui/categories/common/CategoryForm.test.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      onSubmit,
+      mode,
+    };
+
+    return shallow(<CategoryFormBody {...props} />);
+  };
+
+  it('root component renders without crashing', () => {
+    categoryForm = buildCategoryForm();
+    expect(categoryForm.exists()).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/common/CategoryForm.test.js]: toBe

File: test/ui/categories/common/CategoryForm.test.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      expect(code.prop('disabled')).toBe(true);
+    });
+  });
+
+  describe('test buttons and handlers', () => {
+    it('disables submit button while submitting', () => {
+      submitting = true;
+      categoryForm = buildCategoryForm();
+      const submitButton = categoryForm.find('Button');
+      expect(submitButton.prop('disabled')).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/common/CategoryForm.test.js]: toBe

File: test/ui/categories/common/CategoryForm.test.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    });
+
+    it('disables submit button if form is invalid', () => {
+      invalid = true;
+      categoryForm = buildCategoryForm();
+      const submitButton = categoryForm.find('Button');
+      expect(submitButton.prop('disabled')).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/common/CategoryForm.test.js]: toBe

File: test/ui/categories/common/CategoryTreeSelector.test.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+          input={{
+            onChange: ON_CHANGE,
+            value: CATEGORIES[SELECTED_INDEX].code,
+          }}
+        />
+      ));
+    });
+
+    it('renders without crashing', () => {
+      expect(component.exists()).toEqual(true);
```
Comment (fedelocci) [inline test/ui/categories/common/CategoryTreeSelector.test.js]: toBe

File: test/ui/categories/common/CategoryTreeSelectorContainer.test.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  handleExpandCategory: jest.fn().mockReturnValue('handleExpandCategory_result'),
+}));
+
+describe('CategoryTreeSelectorContainer', () => {
+  const dispatchMock = jest.fn();
+
+  describe('mapStateToProps', () => {
+    let props;
+    beforeEach(() => {
+      props = mapStateToProps(STATE_NORMALIZED);
```
Comment (fedelocci) [inline test/ui/categories/common/CategoryTreeSelectorContainer.test.js]: using empty state

File: test/ui/common/form/ActiveLanguagesFields.test.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  let component;
+  beforeEach(() => {
+    component = shallow(<ActiveLanguagesFields
+      activeLanguages={activeLanguages}
+      defaultLanguage={defaultLanguage}
+      onChangeDefaultTitle={onChangeDefaultTitle}
+    />);
+  });
+  it('renders without crashing', () => {
+    expect(component.exists()).toEqual(true);
```
Comment (fedelocci) [inline test/ui/common/form/ActiveLanguagesFields.test.js]: toBe

File: test/ui/common/form/ActiveLanguagesFields.test.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  });
+
+  it('verify if has class ActiveLanguagesFields', () => {
+    expect(component.hasClass('ActiveLanguagesFields')).toEqual(true);
```
Comment (fedelocci) [inline test/ui/common/form/ActiveLanguagesFields.test.js]: toBe

File: src/state/categories/actions.js [hunk 13]
```diff
@@ -37,38 +42,78 @@ export const setCategoryLoaded = categoryCode => ({
...
       dispatch(toggleCategoryExpanded(categoryCode, toExpand));
-      resolve();
     }
+    resolve();
+  });
+
+export const sendPostCategory = categoryData => dispatch =>
+  new Promise((resolve) => {
+    postCategory(categoryData).then((response) => {
+      response.json().then((data) => {
```
Comment (alessioscalici) [inline src/state/categories/actions.js]: Same problem here, if the response is not a json, an error will be raised

File: src/ui/categories/add/AddFormContainer.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    dispatch(fetchLanguages());
+  },
+  onSubmit: data => (dispatch(sendPostCategory(data))),
+  onChangeDefaultTitle: title =>
+    dispatch(change('category', 'code', title.replace(/\W/g, '_').toLowerCase())),
+  onExpandCategory: categoryCode =>
+    dispatch(handleExpandCategory(categoryCode)),
+});
+
+const PagesAddFormContainer = connect(mapStateToProps, mapDispatchToProps)(CategoryForm);
```
Comment (alessioscalici) [inline src/ui/categories/add/AddFormContainer.js]: PagesAddFormContainer doesn't seem so appropriate for a container wrapping CategoryForm  :D

File: src/ui/common/form/ActiveLanguagesFields.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    name: PropTypes.string,
+    isActive: PropTypes.bool,
+    isDefault: PropTypes.bool,
+  })),
+  defaultLanguage: PropTypes.string,
+  onChangeDefaultTitle: PropTypes.func,
+};
+
+ActiveLanguagesFields.defaultProps = {
+  activeLanguages: PropTypes.arrayOf(PropTypes.shape({
```
Comment (alessioscalici) [inline src/ui/common/form/ActiveLanguagesFields.js]: This doesn't look like a default property

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

**Comment by entando-service** [comment]
All is well

**Pull request merged**